components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/AsyncCache.kt

```diff
@@ -0,0 +1,31 @@
+package org.jetbrains.compose.resources
+
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+internal class AsyncCache<K, V> {
+    private val mutex = Mutex()
+    private val cache = mutableMapOf<K, Deferred<V>>()
+
+    suspend fun getOrLoad(key: K, load: suspend () -> V): V = coroutineScope {
+        val deferred = mutex.withLock {
+            var cached = cache[key]
+            if (cached == null || cached.isCancelled) {
+                //LAZY - to free the mutex lock as fast as possible
+                cached = async(start = CoroutineStart.LAZY) { load() }
+                cache[key] = cached
+            }
+            cached
+        }
+        deferred.await()
```

Comment:
We still have an issue when 2 composables starts to load simultaneously:

1. one starts to load
2. the second awaits
3. we exit from the first one
4. the second cancels

Reply:
To solve the issue, we probably need this:
1. use global scope
2. count ref count
3. cancel if ref count is zero